subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/junit/TestClassExecutionEventGenerator.java

```diff
@@ -46,28 +48,41 @@ public TestClassExecutionEventGenerator(TestResultProcessor resultProcessor, IdG
     @Override
     public void testClassStarted(String testClassName) {
         currentTestClass = new DefaultTestClassDescriptor(idGenerator.generateId(), testClassName);
+        testFailed = false;
         resultProcessor.started(currentTestClass, new TestStartEvent(clock.getCurrentTime()));
     }
 
     @Override
-    public void testClassFinished(Throwable failure) {
+    public void testClassFinished(Throwable failure, boolean testClassSkipped) {
         long now = clock.getCurrentTime();
         try {
+            TestResult.ResultType result;
             if (failure != null) {
                 if (currentTests.isEmpty()) {
-                    String testName = testsStarted ? "executionError": "initializationError";
+                    String testName = testsStarted ? "executionError" : "initializationError";
                     DefaultTestDescriptor initializationError = new DefaultTestDescriptor(idGenerator.generateId(), currentTestClass.getClassName(), testName);
                     resultProcessor.started(initializationError, new TestStartEvent(now));
                     resultProcessor.failure(initializationError.getId(), failure);
-                    resultProcessor.completed(initializationError.getId(), new TestCompleteEvent(now));
+                    resultProcessor.completed(initializationError.getId(), new TestCompleteEvent(now, TestResult.ResultType.FAILURE));
                 } else {
                     for (Object test : currentTests) {
                         resultProcessor.failure(test, failure);
-                        resultProcessor.completed(test, new TestCompleteEvent(now));
+                        resultProcessor.completed(test, new TestCompleteEvent(now, TestResult.ResultType.FAILURE));
                     }
                 }
+                // An exception was thrown by the Runner
+                result = TestResult.ResultType.FAILURE;
+            } else if (testFailed) {
+                // One or more tests failed
+                result = TestResult.ResultType.FAILURE;
+            } else if (testClassSkipped) {
+                // The Runner never ran, so we can assume the whole class was skipped
+                result = TestResult.ResultType.SKIPPED;
+            } else {
+                // Even if individual tests were ignored, @Before or @After methods may have run, so set the class status to SUCCESS
+                result = TestResult.ResultType.SUCCESS;
```

Comment:
I doubt this comment. Do you mean the `@BeforeClass` and `@Before` methods are executed even all methods in this class are ignored? This seems not to be the case.

```
public class Test {
    @org.junit.BeforeClass
    public static void before() {
         System.out.println("before class!");
    }

    @org.junit.Before
    public void beforeEach() {
         System.out.println("before!");
    }
    @org.junit.Test
    @org.junit.Ignore
    public void t(){
        throw new RuntimeException();
    }
}
```

Reply:
Maybe the comment is not quite right, what I meant was that to err on the side of caution, even if the class didn't actually execute, mark it as `SUCCESS` in this case, since in the future, only `SKIPPED` status will allow for the optimization where the JVM does not get reforked by the `RestartEveryNTestClassProcessor` or the `ForkingTestClassProcessor`

Reply:
Well, could you please give me an example which can enter this branch block? I failed to think of a situation to let this happen.